src/hotspot/share/runtime/handshake.cpp

```diff
@@ -469,9 +468,9 @@ bool HandshakeState::process_by_self() {
 }
 
 bool HandshakeState::process_self_inner() {
-  while (should_process()) {
+  while (has_operation()) {
     MutexLocker ml(&_lock, Mutex::_no_safepoint_check_flag);
-    HandshakeOperation* op = pop_for_self();
+    HandshakeOperation* op = get_op_for_self();
```

Comment:
I still don't understand why we now do a peek here, and then do the pop below (in remove-op), instead of just doing the pop here?

Reply:
The reason is, from above:
>When poll is armed the handshakee must go to slow path and lock the HandshakeState lock if:
>- 1: There is handshake operation to be processed
>- 2: Some-else is in the middle of processing a handshake operation for this handshakee

This, in practice, means the first pointer of the queue most be non-NULL, and seen as non-NULL, until after we have processed the handshake operation.
By doing:
1: get op
2: do op
  - When handshakee first checks poll, it is armed.
  - Then it checks global poll (safepoint) and the queue state.
  -  If there is a handshake operation to be processed goto slow path.
3: remove op
  - First ptr can be set NULL, which means the handshakee may elide the slow path.
  - The handshakee sees poll armed, but no global poll and nothing in queue.
  - The handshakee may disarm the poll.

By simply removing the OP after we have executed we know the handshakee must go to slow path.

Reply:
Oh, sorry you mean when you are self processing?

It's because of symmetry, everyone do the same, which is far easier to think about.